[DXIL][Analysis] Collect Function properties in Metadata Analysis

bharadwajy opened the pull request:
Basic infrastructure to collect Function properties in Metadata Analysis
- Add a `SmallVector` of entry properties to the metadata information.
- Add a structure to represent function properties. Currently `numthreads` and shader kind properties of shader entry functions are represented.

Commit:
[DXIL][Analysis] Infrastructure to collect Function properties in Metadata Analysis
- Add a map of function-function properties to the metadata information.
- Add a structure to represent various function properties. Currently
  numthreads property of a compute shader entry function is represented.

python3kgae commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +20,16 @@ namespace llvm {
 namespace dxil {
 
+struct FunctionProperties {
+  unsigned NumThreads[3];
```
We might need to put things in a union since not every shader profile has NumThreads.
Something like the PipelinePSVInfo union in PSV https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/DXContainer.h#L238

bharadwajy replied:
> We might need to put things in a union since not every shader profile has NumThreads. Something like the PipelinePSVInfo union in PSV https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/DXContainer.h#L238

Wouldn't the `ShaderStage` field of `struct ModuleMetadataInfo` provide the necessary means for the consumer of the analysis info to identify and distinguish between shaders with functions with `numthreads` property/attribute? Creating a union as part of analysis information to represent the same information seems redundant.

Commit:
Improvements to entry function information collected in Metadata Analysis Pass
- Record shader kind specified for entry function from hlsl.shader
attribute
- Record all entry functions irrespective of the target shader profile
being compute
- Update expected output of tests accordingly

Commit:
clang-format

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -42,6 +83,20 @@ void ModuleMetadataInfo::print(raw_ostream &OS) const {
      << "\n";
   OS << "Validator Version : " << ValidatorVersion.getAsString() << "\n";
+  for (auto MapItem : FunctionPropertyMap) {
+    MapItem.first->getReturnType()->print(OS, false, true);
+    OS << " " << MapItem.first->getName() << "(";
+    FunctionType *FT = MapItem.first->getFunctionType();
+    for (unsigned I = 0, Sz = FT->getNumParams(); I < Sz; ++I) {
+      if (I)
+        OS << ",";
+      FT->getParamType(I)->print(OS);
+    }
+    OS << ")\n";
```
The function name is probably sufficient for debugging purposes, do we really need to go through all of this effort to print a declaration?

bharadwajy replied:
> The function name is probably sufficient for debugging purposes, do we really need to go through all of this effort to print a declaration?

Changed.

bogner reviewed (commented):
Sorry I didn't notice it when the pass itself went in, but we should really have some testing in `llvm/test/Analysis/DXILMetadataAnalysis`, not just piggy backing on the DirectX backend tests.

bharadwajy commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +20,16 @@ namespace llvm {
   Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
   VersionTuple ValidatorVersion{};
-
+  std::unordered_map<Function *, FunctionProperties> FunctionPropertyMap{};
```
> Do we actually need to look up functions in a map for the function properties, or do we just need to be able to iterate through them? It might be nice to wire up the metadata analysis into DXILTranslateMetadata before adding more properties so that we can see how they'll be used.

Function lookup is employed in an implementation I have as a follow on PR that consumes this information in `DXILTranslateMetadata` pass  - see [here](https://github.com/bharadwajy/llvm-project/blob/0da45710ee0bfbc505155ff61929b9c1ed83e806/llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp#L69C4-L70C65).

bharadwajy replied:
Changed to `MapVector`.

bharadwajy commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +20,16 @@ namespace llvm {
 
+struct FunctionProperties {
+  unsigned NumThreads[3];
+};
```
> Are these function properties or entry point properties? What else do we expect to need in this struct?

The intent is to provide a generic structure to hold any function properties (if/as needed) - but am not opposed to renaming it to `EntryProperties` if other properties of other functions are not collected or of use.

bharadwajy replied:
> > Are these function properties or entry point properties? What else do we expect to need in this struct?
> 
> The intent is to provide a generic structure to hold any function properties (if/as needed) - but am not opposed to renaming it to `EntryProperties` if other properties of other functions are not collected or of use.

Renamed.

Commit:
Address PR feedback

Move testing of DXILMetadata Analysis to
llvm/test/Analysis/DXILMetadataAnalysis from
llvm/test/CodeGen/DirectX/Metadata and add a new test

bharadwajy commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +20,16 @@ namespace llvm {
 namespace dxil {
 
+struct FunctionProperties {
+  unsigned NumThreads[3];
```
> Would it be more usable to just specify the three numbers individually? I doubt that iterating through these is particularly meaningful.

Changed as suggested.

bharadwajy commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,6 +38,42 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+        report_fatal_error(Twine(F.getName()) +
+                               ": Invalid numthreads specified",
+                           /* gen_crash_diag */ false);
+      }
```
> This should just be an assert - we expect the IR to be valid at this point. No need for `report_fatal_error` here.

Changed.

bharadwajy commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,6 +38,42 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+               "Failed to parse numthreads components as integer values");
+        unsigned &Num = std::get<1>(It);
+        Num = V.getLimitedValue();
+      }
```
> I don't think the use of `zip` here makes this clearer than doing it the obvious way. Also the code inside an assert isn't executed in release builds, so it really shouldn't have side effects.

Changed.

bharadwajy commented:
> Sorry I didn't notice it when the pass itself went in, but we should really have some testing in `llvm/test/Analysis/DXILMetadataAnalysis`, not just piggy backing on the DirectX backend tests.

Moved testing of DXIL Metadata Analysis out of DirectX backend tests to `llvm/test/Analysis/DXILMetadataAnalysis` and added an additional test.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +21,20 @@ namespace llvm {
   Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
   VersionTuple ValidatorVersion{};
-
+  MapVector<Function *, EntryProperties> EntryPropertyMap;
```
Why a MapVector? Is the insertion order of these functions interesting to preserve for some reason, or is this just here so that it's easy to make the order that we print them in deterministic?

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +38,54 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+      NumThreadsStr.split(NumThreadsVec, ',');
+      assert(NumThreadsVec.size() == 3 && "Invalid numthreads specified");
+      // Read in the three component values of numthreads
+      if (!llvm::to_integer(NumThreadsVec[0], EFP.NumThreadsX, 10))
+        assert(false && "Failed to parse X component of numthreads");
```
`if (x) assert(false)` is generally discouraged. It doesn't make much difference, but it's probably better 
```c++
    [[maybe_unused]] bool Success =
        llvm::to_integer(NumThreadsVec[0], EFP.NumThreadsX, 10);
    assert(Success&& "Failed to parse X component of numthreads");
```

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +38,54 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+      if (!llvm::to_integer(NumThreadsVec[2], EFP.NumThreadsZ, 10))
+        assert(false && "Failed to parse Z component of numthreads");
+    }
+    MMDAI.EntryPropertyMap.insert(std::make_pair(std::addressof(F), EFP));
```
Why use `std::addressof(F)` here rather than just `&F`? Generally you only need `std::addressof` in certain template contexts

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +38,54 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+  OS << "Target Shader Stage : " << Triple::getEnvironmentTypeName(ShaderStage)
      << "\n";
   OS << "Validator Version : " << ValidatorVersion.getAsString() << "\n";
+  for (auto MapItem : EntryPropertyMap) {
```
We can use C++17 so you could use structured bindings to clean this up slightly if you want. In any case this should probably be `const auto &` to avoid copying the items unnecessarily

bogner reviewed (commented):
A few more questions and comments. We should have a test with multiple entry points as well.

bharadwajy commented:
> A few more questions and comments. We should have a test with multiple entry points as well.

I've already added one - `llvm/test/Analysis/DXILMetadataAnalysis/entry-properties.ll`.

Commit:
Address latest PR feedback

Collect Entry properties in a SmallVector instead of in a Mapvector.

Use structured binding declaration for printing analysis info.

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -9,6 +9,8 @@
 #define LLVM_ANALYSIS_DXILMETADATA_H
 
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/Function.h"
```
A forward declaration should be sufficient for `Function`

bogner commented on llvm/include/llvm/Analysis/DXILMetadataAnalysis.h:
```diff
@@ -19,12 +21,23 @@ namespace llvm {
   Triple::EnvironmentType ShaderStage = Triple::UnknownEnvironment;
   VersionTuple ValidatorVersion{};
-
+  SmallVector<EntryProperties, 4> EntryPropertyVec{};
```
Unless there's a specific argument for 4, it's better to use `SmallVector<EntryProperties>` which will automatically deduce a reasonable size

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -8,11 +8,16 @@
 #include "llvm/IR/Module.h"
 #include "llvm/InitializePasses.h"
+#include "llvm/Support/ErrorHandling.h"
+#include <memory>
+#include <utility>
```
I don't think `<memory>` or `<utility>` are used any more

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +38,52 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+      NumThreadsStr.split(NumThreadsVec, ',');
+      assert(NumThreadsVec.size() == 3 && "Invalid numthreads specified");
+      // Read in the three component values of numthreads
+      bool Success = llvm::to_integer(NumThreadsVec[0], EFP.NumThreadsX, 10);
```
You need a `[[maybe_unused]]` or a `(void)Success` so that this doesn't trigger warnings in no-asserts builds

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +38,52 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+  OS << "Target Shader Stage : " << Triple::getEnvironmentTypeName(ShaderStage)
      << "\n";
   OS << "Validator Version : " << ValidatorVersion.getAsString() << "\n";
+  for (const auto [Ent, SS, NX, NY, NZ] : EntryPropertyVec) {
```
I think this should be `const auto &[Ent, SS, NX, NY, NZ]` to avoid a copy. This is also starting to get a bit unwieldy with this many members in the structured binding.

bogner commented:
Also make sure to update the PR description - it still mentions a map.

Commit:
A few more changes from PR feedback.

Commit:
clang-format change

bogner commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -33,15 +36,54 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+  OS << "Target Shader Stage : " << Triple::getEnvironmentTypeName(ShaderStage)
      << "\n";
   OS << "Validator Version : " << ValidatorVersion.getAsString() << "\n";
+  for (const auto EP : EntryPropertyVec) {
```
I've already pointed this out a couple of times (https://github.com/llvm/llvm-project/pull/105728#discussion_r1739308959, https://github.com/llvm/llvm-project/pull/105728#discussion_r1739469201), but please don't copy the value here. It's inefficient, and a const reference will do just fine.

bharadwajy replied:
> I've already pointed this out a couple of times ([#105728 (comment)](https://github.com/llvm/llvm-project/pull/105728#discussion_r1739308959), [#105728 (comment)](https://github.com/llvm/llvm-project/pull/105728#discussion_r1739469201)), but please don't copy the value here. It's inefficient, and a const reference will do just fine.

Changed.

Commit:
Change inadvertently mis-typed for-loop item declaration to reference type.
